casper/src/test/resources/PoSTest.rho

```diff
@@ -3,32 +3,290 @@
-        RhoSpecCh,
-        PosCh,
-        setup,
-        test,
-        test_different_deployer,
-        ackCh in {
+          contract setup(retCh) = {
+            retCh!({})
+          } |
 
-      rl!(`rho:id:6wnujzcraztjfg941skrtbdkdgbko8nuaqihuhn15s66oz8ro5gwbb`, *RhoSpecCh) |
-      rl!(`rho:id:cnec3pa8prp4out3yc8facon6grm3xbsotpd4ckjfx8ghuw77xadzt`, *PosCh) |
-      for(@(_, RhoSpec) <- RhoSpecCh; @(_, PoS) <- PosCh) {
+          contract prepareUser(@pk, @vaultAmount, retCh) = {
+            new setDeployData(`rho:test:deploy:set`), identitySet, revAddrCh, vaultCh in {
+              setDeployData!("userId", pk.hexToBytes(), *identitySet) |
+              revAddressOps!("fromPublicKey", pk.hexToBytes(), *revAddrCh) |
+              for (_ <- identitySet; @revAddress <- revAddrCh) {
+                @RevVault!("findOrCreateGenesisVault", revAddress, vaultAmount, *vaultCh) |
+                for (@(true, vault) <- vaultCh) {
+                  retCh!(pk.hexToBytes(), revAddress, vault)
+                }
+              }
+            }
+          } |
+
+          contract test_make_pos_succeeds(rhoSpec, _, ackCh) = {
+            new retCh in {
+              @PoS!("getBonds", *retCh) |
+              for ( @bonds <- retCh) {
+                rhoSpec!("assert",
+                  bonds.size() >= 0,
+                  "some bonds map should be received",
+                  *ackCh
+                )
+              }
+            }
+          } |
+
+          contract test_bonding_succeeds(rhoSpec, _, ackCh) = {
+            new setupCh, retCh, initialPosBalanceCh, finalPosBalanceCh,
+                initialUserBalanceCh, finalUserBalanceCh in {
+              prepareUser!("1111111111111111111111111111111111111111111111111111111111111111", 10000, *setupCh) |
+              for (@user1PubKey, _, @user1Vault <- setupCh) {
+                @posVault!("balance", *initialPosBalanceCh) |
+                @user1Vault!("balance", *initialUserBalanceCh) |
+                for(@initialPosBalance <- initialPosBalanceCh;
+                    @initialUserBalance <- initialUserBalanceCh) {
+                  @PoS!("bond", 100, *retCh) |
+                  for ( @(true, _) <- retCh) {
+                    @PoS!("getBonds", *retCh) |
+                    @posVault!("balance", *finalPosBalanceCh) |
+                    @user1Vault!("balance", *finalUserBalanceCh) |
+                    for(@finalPosBalance <- finalPosBalanceCh;
+                        @finalUserBalance <- finalUserBalanceCh;
+                        @bonds <- retCh) {
+                      stdlog!("info", {"finalPosBalance": finalPosBalance, "finalUserBalance":finalUserBalance, "bonds":bonds}) |
+                      match bonds.get(user1PubKey) {
+                        [bondAmount, _] => {
+                          rhoSpec!("assertMany",
+                            [
+                              ((100, "==", bondAmount), "the new bond is expected in the final map"),
```

Comment:
That's a lot of nesting that can be avoided by ways similar to `assertBalances` in RevVaultTest.rho (we pass a map of balances to assert, we pass a "now's your turn" channel as the first parameter and wait till triggered to do the assertions